memory read is now completely stable even on intermittently slow targets

timsifive opened the pull request:
The downloaded program now post-increments, and there's no longer an attempt to read the current address from the target. This made it easier to fix the problem where at the start of the loop the current address was already read (in regular entry) or has not yet been read (when the first round through the loop encountered busy more than once, or busy was encountered at least once later on).

Fixed a few math errors regarding addresses/indexes.

Also:
* Removed unnecessary reset in increase_ac_busy_delay().
* Removed various unnecessary newlines in log messages.

This version passes a stress test (simultaneously run 4 spikes with random delays added in, covering 32- and 64-bit, as well as single and dual core), and it also passes against the hardware I have access to.

timsifive committed:
Remove unnecessary newlines.

timsifive committed:
Don't reset DMI when an abstract command is busy.

timsifive committed:
Fix off-by-3 error on 64-bit targets.

This caused everything to fall apart when debugging slow 64-bit targets.

timsifive committed:
Remove redundant code.

timsifive committed:
Fix off-by-one error.

timsifive committed:
Finally nailed memory read on slow targets

The downloaded program now post-increments, and there's no longer an
attempt to read the current address from the target. This made it easier
to fix the problem where at the start of the loop the current address
was already read (in regular entry) or has not yet been read (when the
first round through the loop encountered busy more than once, or busy
was encountered at least once later on).

palmer-dabbelt reviewed (changes requested):
I always find the read_memory function hard to read.  Does it still respect the constraint that it only reads from each address once?  IIRC that's why we had this off-by-one oddness here.

timsifive commented:
If the timing is right, and the loop doesn't encounter cmderr=busy, then every address is read exactly once. When cmderr=busy an entire block is read again, which means the addresses at the start of that block may be read more than once.

I will fix that, although I don't know of a good way to add a test for it. I'll also add a comment mentioning that behavior is important.

palmer-dabbelt commented:
It's probably more work to test for than just to keep sane manually.  The constraint is for memory-mapped register reads, which sometimes go through this path, so we could always test it with a special device in Spike.

timsifive commented:
Note that for reads of a single word, the read is never executed twice.

timsifive committed:
Ensure read_memory() only reads each address once.

Previously it might read an address multiple times if an abstract
command took longer to execute than expected.

The new implementations reads from the target how far it has gotten
along reading memory, and resumes from there if cmderr=busy.

This ended up being a bigger change than I envisioned, but in the end it
deleted more lines than it added, so I'm happy. :-)

timsifive commented:
Fixed.
